module/VuFindSearch/src/VuFindSearch/Backend/WorldCat2/Connector.php

```diff
@@ -0,0 +1,211 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  WorldCat
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+
+namespace VuFindSearch\Backend\WorldCat2;
+
+use Exception;
+use Laminas\Http\Client\Exception\RuntimeException as ExceptionRuntimeException;
+use Laminas\Http\Exception\InvalidArgumentException;
+use Laminas\Http\Exception\RuntimeException;
+use Laminas\Http\Response;
+use Laminas\Log\LoggerAwareInterface;
+use Laminas\Session\Container;
+use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
+use League\OAuth2\Client\Provider\GenericProvider;
+use VuFind\Log\LoggerAwareTrait;
+use VuFindSearch\ParamBag;
+
+/**
+ * Class for accessing OCLC WorldCat search API v2.
+ *
+ * @category VuFind
+ * @package  WorldCat
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class Connector implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    /**
+     * API base URL
+     *
+     * @var string
+     */
+    protected $baseUrl;
+
+    /**
+     * Constructor
+     *
+     * @param \Laminas\Http\Client $client       An HTTP client object
+     * @param GenericProvider      $authProvider OAuth2 provider
+     * @param Container            $session      Session container for persisting data
+     * @param array                $options      Additional config settings
+     */
+    public function __construct(
+        protected \Laminas\Http\Client $client,
+        protected GenericProvider $authProvider,
+        protected Container $session,
+        protected array $options = []
+    ) {
+        if (empty($options['base_url'])) {
+            throw new \Exception('base_url setting is required');
+        }
+        $this->baseUrl = $options['base_url'];
+    }
+
+    /**
+     * Get an OAuth2 token.
+     *
+     * @return string
+     * @throws IdentityProviderException
+     */
+    protected function getToken(): string
+    {
+        return $this->authProvider->getAccessToken('client_credentials', ['scope' => 'wcapi'])->getToken();
+    }
+
+    /**
+     * Make an API call.
+     *
+     * @param string $path Path to query (including GET parameters)
+     *
+     * @return Response
+     * @throws IdentityProviderException
+     * @throws InvalidArgumentException
+     * @throws RuntimeException
+     * @throws ExceptionRuntimeException
+     */
+    protected function makeApiCall(string $path): Response
+    {
+        if (!isset($this->session->token)) {
+            $this->session->token = $this->getToken();
```

Comment:
If the auth credentials are wrong (or never set) in the .ini, this throws a not-very-helpful exception from the oauth2 library deep within `getToken`.  

![image](https://github.com/user-attachments/assets/e9e295c6-d84c-4136-9e85-1be277a8cd81)

Worth catching somehow and faking a 401 so it can be handled a few lines down.

Reply:
I'm not sure I understand what you mean. If there is no token in the session, we try to get a fresh token. The logic a few lines down attempts to get a fresh token if the one in the session has expired. If failing to get the token the first time led to us simulating a 401, we'd just end up repeating the same call that failed in the first place. I don't think that would help! Am I misunderstanding something?

Reply:
Ok yeah "simulating" a 401 response is not a good idea.  My point (badly stated) was that when I have an actual authentication problem (wrong password), there is an uncaught exception thrown, not a response with a 401.  So the code handling the 401 never gets triggered.  May be different if the reason for the 401 is actually an expired token instead of a bad password.

Reply:
My point is that this code is about dealing with a 401 coming from WorldCat, not with sending a 401 to the end user. But I'm not sure that it's appropriate to send a 401 to the VuFind user when this happens -- it's more of a 500 situation, since the server is misconfigured...

Reply:
Ok I think that's a fair point.  I hate how opaque the error message is for the developer, but ultimately it's a 500 which as you say is correct.

Reply:
Yeah, it's not ideal. If this turns out to be a recurring support problem/source of frustration, we might at least be able to wrap the low-level exception in a higher-level exception... but that could also make things even more confusing if the failures differ from what we expect them to be. I'll leave it alone for now but we can revisit if needed!